include/cupla/math/Common.hpp

```diff
@@ -0,0 +1,114 @@
+/* Copyright 2020 Rene Widera
+ *
+ * This file is part of cupla.
+ *
+ * cupla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * cupla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with cupla.
+ * If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+
+#pragma once
+
+#include "cupla/types.hpp"
+
+#include <alpaka/core/Concepts.hpp>
+
+#include <type_traits>
+
+namespace cupla
+{
+inline namespace CUPLA_ACCELERATOR_NAMESPACE
+{
+inline namespace device
+{
+inline namespace math
+{
+namespace detail
+{
+    /** Get the concept implementation of the current accelerator
+     *
+     * @tparam T_Concept alpaka concept
+     * @return implementation of the concept
+     */
+    template< typename T_Concept >
+    ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE auto getConcept()
+    -> alpaka::concepts::ImplementationBase<
+        T_Concept,
+        Acc
+    >
+    {
+        using AccThreadSeqMathConcept = alpaka::concepts::ImplementationBase<
+            T_Concept,
+            AccThreadSeq
+        >;
+        using AccMathConcept = alpaka::concepts::ImplementationBase<
+            T_Concept,
+            Acc
+        >;
+        // cupla Acc and AccThreadSeq should use the same math concept implementation
+        static_assert(
+            std::is_same<
+                AccMathConcept,
+                AccThreadSeqMathConcept
+            >::value,
+            "The math concept implementation for the type 'Acc' and 'AccThreadSeq' must be equal"
+        );
+        return AccMathConcept{};
+    }
+} // namespace detail
+
+#define CUPLA_UNARY_MATH_FN(functionName, alpakaMathConcept)                   \
+    /**                                                                        \
+     * @tparam T_Type argument type                                            \
+     * @param arg input argument                                               \
+     */                                                                        \
+    template< typename T_Type >                                                \
+    ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE auto functionName(                     \
```

Comment:
note: the function attribute must be `ALPAKA_FN_HOST_ACC` else it will not fit the definition in alpaka.